### PR TITLE
Support Azure in Snowflake

### PR DESCRIPTION
**Pull request by alexrenz**
Add support for running on Azure inside Snowflake.

### CODE REVIEW

File: src/snowflake/mod.rs [hunk 1]
```diff
@@ -92,6 +105,93 @@ impl object_store::CredentialProvider for S3StageCredentialProvider {
...
+        let mut locked = self.cached.lock().await;
+
+        match locked.as_ref() {
+            Some(creds) => {
+                if matches!(creds.as_ref(), AzureCredential::SASToken(pairs) if *pairs == new_pairs) {
+                    return Ok(Arc::clone(creds));
+                }
+            }
+            _ => {}
+        }
```
Comment (alexrenz) [inline src/snowflake/mod.rs:187]: @andrebsguedes: I adapted this from the S3 code + your variant, but don't follow why we have this caching here. 

Speaking about the S3 variant: the request to Snowflake in `current_upload_info` (which happens before the cache check) should be more expensive than  creating an `object_store::aws::AwsCredential` (which happens after), right? I also don't have a good understanding of when object_store calls `get_credential`, maybe that is my problem.

File: src/snowflake/kms.rs [hunk 2]
```diff
@@ -193,3 +177,188 @@ impl CryptoMaterialProvider for SnowflakeStageKms {
...
+            query_id: encryption_material.query_id.clone(),
+            key_size: "128".to_string()
+        };
+        let master_key = Key::from_base64(&encryption_material.query_stage_master_key)
+            .map_err(ErrorKind::MaterialDecode)?;
+
+        let scheme = self.config.crypto_scheme;
+        let material = ContentCryptoMaterial::generate(scheme);
+        let encrypted_cek = material.cek.clone().encrypt_aes_128_ecb(&master_key)
+            .map_err(ErrorKind::MaterialCrypt)?;
```
Comment (alexrenz) [inline src/snowflake/kms.rs:246]: @andrebsguedes I am confused by this. The encryption data says that the key wrapping uses `AES_CBC_256` but the master keys that I am seeing are 128bit. Do you know more about this?

File: src/snowflake/client.rs [hunk 3]
```diff
@@ -614,6 +661,30 @@ impl SnowflakeClient {
         }).await?;
         Ok(stage_info)
     }
+    pub(crate) async fn get_master_key(
+        &self,
+        query_id: String,
+        path: &str,
+        stage: &str,
+        keyring: &Cache<String, Key>,
```
Comment (andrebsguedes) [inline src/snowflake/client.rs]: I really like the DRY-_ness_ of this function and we should keep it, but I think it is a little awkward from the encapsulation perspective to have a method in `SnowflakeClient` that needs to know about the `keyring` type.

My suggestion would be to move this function to `src/snowflake/kms.rs` and instead pass the `SnowflakeClient` as an argument.

If a free function is not elegant enough we could wrap the `Cache` in a new `Keyring` type that also has an `Arc<SnowflakeClient>` and does all of this transparently within some new `get` method, but I am not sure it is worth it to go as far as this (how knows if fetching the `master_key` in GCP requires performing an smoke signal to the Snowflake servers or something :upside_down_face: :laughing:)

File: src/snowflake/kms.rs (hunk 2 above)
Comment (andrebsguedes) [reply src/snowflake/kms.rs:246]: Yes all keys from Snowflake are currently 128 bits but in order to keep everything compatible with the Azure clients they set the `wrapped_content_key.algorithm` to `AES_CBC_256` because that is the only AES CBC variant the clients support (even though it would fail anyway :upside_down_face: ). So we just do exactly what they do

File: src/snowflake/kms.rs [hunk 4]
```diff
@@ -193,3 +177,184 @@ impl CryptoMaterialProvider for SnowflakeStageKms {
...
+        };
+        
+        Ok(content_material)
+    }
+}
+
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "PascalCase")]
+struct EncryptionData {
```
Comment (andrebsguedes) [inline src/snowflake/kms.rs:337]: I really like the fully defined structs here for deriving `Serialize` and `Deserialize`, this way we prevent typos and perform the field validation in a single go when deserializing.

File: src/snowflake/mod.rs [hunk 5]
```diff
@@ -301,10 +401,69 @@ pub(crate) async fn build_store_for_snowflake_stage(
...
+
+
+            let provider = AzureStageCredentialProvider::new(&config.stage, client.clone());
+
+            match info.stage_info.test_endpoint.as_deref() {
+                Some(_) => {
+                    unimplemented!("test endpoint for azure blob storage is not supported");
+                }
+                None => {}
+            }
```
Comment (andrebsguedes) [inline src/snowflake/mod.rs]: That could be written like this (because you don't need the value):
```suggestion
            if info.stage_info.test_endpoint.is_some() {
                unimplemented!("test endpoint for azure blob storage is not supported");
            }
```

File: src/util.rs [hunk 6]
```diff
@@ -485,3 +486,10 @@ where
     let de = &mut serde_json::Deserializer::from_str(v);
     serde_path_to_error::deserialize(de)
 }
+
+pub(crate) fn required_attribute<'a>(key: &'static str, attr: &'a Attributes) -> Result<&'a str, Error> {
```
Comment (andrebsguedes) [inline src/util.rs]: [minor nit] Good call on moving this to here, I would just swap the order of the arguments because `required_attribute(&attr, "x-amz-matdesc")` looks more like `attr["x-amz-matdesc"]` and thus more idiomatic to me

File: src/snowflake/mod.rs [hunk 7]
```diff
@@ -92,6 +105,93 @@ impl object_store::CredentialProvider for S3StageCredentialProvider {
...
+
+        let new_creds = info.stage_info.creds.as_azure()
+            .map_err(|e| object_store::Error::Generic {
+                store: "MicrosoftAzure",
+                source: e.into()
+            })?;
+        
+
+        let token_bytes = new_creds.azure_sas_token.trim_start_matches('?').as_bytes();
+        let new_pairs = url::form_urlencoded::parse(token_bytes)
```
Comment (andrebsguedes) [inline src/snowflake/mod.rs:174]: I was not aware that `url` re-exports `form_urlencoded`! That looks much cleaner than the hack on my version.

File: src/snowflake/mod.rs (hunk 1 above)
Comment (andrebsguedes) [reply src/snowflake/mod.rs:187]: Yeah, you are right that this caching here is very confusing and my Azure version of this was pretty dumb to be honest, let me try to explain the rationale:
- object_store calls this for every operation, even multiple times in some cases, so we want this to be fast
- `current_upload_info` is a cached call to `fetch_upload_info` which means that most of the time calling it should just be a few atomic operations (to read the upload_info cache and perform the `Arc::clone`) and thus fast.
- So we call `current_upload_info` but instead of cloning all the strings to build `AwsCredential` we cache it so that we can `Arc::clone` it most of the time

Some critical analysis after looking at my S3 code again: 
- In reality even if everything goes great the extra contention from the lock may be way worse than the allocations for constructing `AwsCredential`.
- We would have to employ way more clever tricks with `arc_swap` and use the address of the `current_upload_info` result to detect change instead of checking the contents

TLDR: I think this caching is not worth it neither in S3 or Azure and we should drop it for now

**Review by andrebsguedes — COMMENTED** [review]
Looks awesome! Even more so when we take into account this is your first PR in the project.

Some minor comments:

File: src/snowflake/client.rs (hunk 3 above)
Comment (alexrenz) [reply src/snowflake/client.rs]: Agreed. Made it a free function 👍

File: src/snowflake/mod.rs (hunk 5 above)
Comment (alexrenz) [reply src/snowflake/mod.rs]: uuhhh an `if` in Rust :D A rare sight!  Done. Thanks for the suggestion!

File: src/util.rs (hunk 6 above)
Comment (alexrenz) [reply src/util.rs]: Done 👍

File: src/snowflake/mod.rs (hunk 1 above)
Comment (alexrenz) [reply src/snowflake/mod.rs:187]: Agree to drop it. Let's do that in another PR. So I will keep this as is here

File: Cargo.toml [hunk 8]
```diff
@@ -44,6 +44,7 @@ reqwest = { version = "0.12", default-features = false, features = ["rustls-tls"
 # object_store = { version = "0.10.1", features = ["azure", "aws"] }
 # Pinned to a specific commit while waiting for upstream
 object_store = { git = "https://github.com/andrebsguedes/arrow-rs.git", tag = "v0.10.2-beta1", features = ["azure", "aws", "experimental-azure-list-offset", "experimental-arbitrary-list-prefix"] }
+hickory-resolver = "0.24"
```
Comment (alexrenz) [inline Cargo.toml:47]: @andrebsguedes Is there a way to say sth like "use whatever version reqwest is using"?

Comment (andrebsguedes) [reply Cargo.toml:47]: Unfortunately not. We can add a test later that at least breaks if the versions diverge but for this I don't think it is even needed as reqwest does not care about the version of a custom resolver

**Review by andrebsguedes — APPROVED** [review]
Looks great! I still have to go through the raicode tests but it is very unlikely that the review there will cause any changes at this level so I will approve this one already.
